File: java/end-to-end-applications/food-ordering/app/restate-app/src/main/java/dev/restate/sdk/examples/OrderStatusService.java
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate examples,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

package dev.restate.sdk.examples;

import dev.restate.sdk.ObjectContext;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.VirtualObject;
import dev.restate.sdk.examples.types.StatusEnum;
import dev.restate.sdk.common.StateKey;
import dev.restate.sdk.common.TerminalException;

@VirtualObject
public class OrderStatusService {

  private static final StateKey<StatusEnum> ORDER_STATUS =
      StateKey.of("order-status", StatusEnum.class);
  private static final StateKey<Long> ORDER_ETA = StateKey.of("order-eta", Long.TYPE);

  public static class OrderStatus {
    private final StatusEnum status;
    private final long eta;

    public OrderStatus(StatusEnum status, long eta) {
      this.status = status;
      this.eta = eta;
    }

    public long getEta() {
      return eta;
    }

    public StatusEnum getStatus() {
      return status;
    }
  }

  /** Gets called by the webUI frontend to display the status of an order. */
  @Handler
  public OrderStatus get(ObjectContext ctx) throws TerminalException {
    var status = ctx.get(ORDER_STATUS).orElse(StatusEnum.NEW);
    var eta = ctx.get(ORDER_ETA).orElse(-1L);
    return new OrderStatus(status, eta);
  }

  @Handler
  public void setStatus(ObjectContext ctx, StatusEnum statusEnum) throws TerminalException {
    ctx.set(ORDER_STATUS, statusEnum);
  }

  @Handler
  public void setETA(ObjectContext ctx, long eta) throws TerminalException {
    ctx.set(ORDER_ETA, eta);
  }
}

File: end-to-end-applications/kotlin/food-ordering/app/restate-app/src/main/kotlin/dev/restate/sdk/examples/OrderStatusService.kt
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate examples,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */
package dev.restate.sdk.examples

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.VirtualObject
import dev.restate.sdk.kotlin.KtStateKey
import dev.restate.sdk.kotlin.ObjectContext

@VirtualObject
class OrderStatusService {
  companion object {
    private val ORDER_STATUS = KtStateKey.json<StatusEnum>("order-status")
    private val ORDER_ETA = KtStateKey.json<Long>("order-eta")
  }

  /** Gets called by the webUI frontend to display the status of an order. */
  @Handler
  suspend fun get(ctx: ObjectContext): OrderStatus {
    return OrderStatus(ctx.get(ORDER_STATUS) ?: StatusEnum.NEW, ctx.get(ORDER_ETA) ?: -1L)
  }

  @Handler
  suspend fun setStatus(ctx: ObjectContext, statusEnum: StatusEnum) {
    ctx.set(ORDER_STATUS, statusEnum)
  }

  @Handler
  suspend fun setETA(ctx: ObjectContext, eta: Long) {
    ctx.set(ORDER_ETA, eta)
  }
}
